library/extractor/src/main/java/com/google/android/exoplayer2/extractor/ogg/OpusReader.java

```diff
@@ -69,13 +76,41 @@ protected boolean readHeaders(ParsableByteArray packet, long position, SetupData
               .setSampleRate(OpusUtil.SAMPLE_RATE)
               .setInitializationData(initializationData)
               .build();
+
       headerRead = true;
       return true;
     } else {
       checkNotNull(setupData.format); // Has been set when the header was read.
-      boolean headerPacket = packet.readInt() == OPUS_CODE;
-      packet.setPosition(0);
-      return headerPacket;
+
+      // Handle the comments frame if we are working with one.
+      boolean isComment;
+
+      if (packet.bytesLeft() < OPUS_TAG_SIGNATURE.length) {
+        isComment = false;
+      } else {
+        byte[] header = new byte[OPUS_TAG_SIGNATURE.length];
+        packet.readBytes(header, 0, OPUS_TAG_SIGNATURE.length);
+        isComment = Arrays.equals(header, OPUS_TAG_SIGNATURE);
+      }
+
+      if (isComment) {
+        VorbisUtil.CommentHeader commentHeader = VorbisUtil.readVorbisCommentHeader(
+            packet, false, false);
+
+        @Nullable Metadata vorbisMetadata = VorbisUtil.buildMetadata(
+            Arrays.asList(commentHeader.comments)
+        );
+
+        setupData.format = setupData.format
+            .buildUpon()
+            .setMetadata(vorbisMetadata)
+            .build();
+
+        return true;
+      } else {
+        packet.setPosition(0);
```

Comment:
I think this is an accidental side-effect of me trying to shoehorn comment header parsing into this routine.

More or less, the program would normally do the following:
- Does this header begin with "O p u s"?
    - If yes, then keep going, we have more headers to consume
    - If no, then stop here.

Since `OpusReader` does this indiscriminately to all frames, it has to rewind with this method or it would corrupt the packet. However, this is no longer needed as we now parse this frame.

According to RFC7845, the comment header is "required", so perhaps we should two one of the following:
- Throw an error if the second packet does not have the comment signature, as we expect it to [I'm in favor of this one]
- Just stop parsing if the packet is invalid.

Reply:
Actually, after making that change, it seems like changing it so that it does something more "sane" causes the extractor test to break. 

Iteems like the semantics of `readHeaders` are the following:
- `return true` -> We just consumed the header, keep going
- `return false` -> This is not a header, stop now

This would explain why it breaks. I'll keep the current process of checking the ID, but I've simplified it and eliminated the possible bug that might occur with setting the position to 0.

Reply:
> The old code would call setPosition(0) both when it found a comment header and when it didn't. This new code only calls it when it doesn't find a comment header.

This seems correct (see point 1).

> The documentation of StreamReader#readHeaders isn't very clear what's expected here :(

Yes, I agree the documentation is not very good :(

1. As long as `OpusReader.readHeaders` returns true (i.e. the packet contains headers), it doesn't matter where the position of `packet` is at the end of the method execution, because `oggPacket` will be re-populated with the next packet in `StreamReader.readHeaders`. But if `OpusReader.readHeaders` returns false, then the position of `packet` must be set to 0 because it will be re-read in `StreamReader.readPayload`. The method name `OpusReader.readHeaders` is confusing because in some cases where it finds a header, it doesn't read it entirely and resets the position to 0.

2.  I think it depends on how you see it. Arguably, the parameter name `packet` implies that the `ParsableByteArray` contains the packet and the packet only (and therefore that the packet starts at position 0). I don't think there is any bug here because, from what I see in the code, `packet` should always only contain the packet. SGTM to remove the assumption that the packet starts at 0 but the code as it is now also seems fine to me.

Reply:
I've tried to resolve this with b919161. To prevent any issues from occuring with `setPosition(0)`, I instead get the initial position beforehand with `getPosition` and rewind to that if the comment header fails to parse.